bundler/helpers/lib/functions/version_resolver.rb

```diff
@@ -0,0 +1,206 @@
+module Functions
+  class VersionResolver
+    GEM_NOT_FOUND_ERROR_REGEX = /locked to (?<name>[^\s]+) \(/.freeze
+
+    attr_reader :dependency_name, :dependency_requirements,
+                :gemfile_name, :lockfile_name,
+                :dir, :credentials
+
+    def initialize(dependency_name:, dependency_requirements:,
+                   gemfile_name:, lockfile_name:, using_bundler_2:,
+                   dir:, credentials:)
+      @dependency_name = dependency_name
+      @dependency_requirements = dependency_requirements
+      @gemfile_name = gemfile_name
+      @lockfile_name = lockfile_name
+      @using_bundler_2 = using_bundler_2
+      @dir = dir
+      @credentials = @credentials
+    end
+
+    def version_details
+      setup_bundler
+
+      dep = dependency_from_definition
+
+      # TODO: Rewrite this
+      #
+      # If the dependency wasn't found in the definition, but *is*
+      # included in a gemspec, it's because the Gemfile didn't import
+      # the gemspec. This is unusual, but the correct behaviour if/when
+      # it happens is to behave as if the repo was gemspec-only.
+      if dep.nil? && dependency_requirements.any?
+        return "latest"
+      end
```

Comment:
This is a bad signal to pass back, I took a shot at just passing back nil and then examining `dependency_requirements.any?` in the calling code, but it broke several tests surprisingly.

Will review now the rest of the refactor is relatively sane.